### DISCUSSION TITLE
Z2M 2.0.0: `USB adapter discovery error (No valid USB adapter found)`

### DISCUSSION THREAD

**Discussion by Koenkk** [Koenkk/zigbee2mqtt#24199] [General]
Locked due to https://github.com/Koenkk/zigbee2mqtt/discussions/24199#discussioncomment-10964035, follow-up: https://github.com/Koenkk/zigbee2mqtt/discussions/24364

**TL;DR**

Add `adapter` (likely `zstack`, see [docs](https://github.com/Koenkk/zigbee2mqtt.io/blob/fix/adapter-discovery/docs/guide/configuration/adapter-settings.md#basic-configuration) for more info) to your Zigbee2MQTT `configuration.yaml`, example:

```
serial:
  port: /dev/ttyACM0
  adapter: zstack # <-- Add this line
```

**What changed?**

https://github.com/Koenkk/zigbee-herdsman/pull/1197 brings improvements to the adapter discovery feature. This feature allows Zigbee2MQTT to automatically detect your adapter (preventing the need for a `serial` section in `configuration.yaml`). 

In case Zigbee2MQTT could not automatically detect the type of adapter, it would default to `zstack`. This gives issues to people using other adapters. When not specifying the `adapter` and when using e.g. an EZSP adapter, Zigbee2MQTT will fail to start with `Error: SRSP - SYS - ping after 6000ms` instead of a clear error message. This behaviour has now been changed. Starting from 1.41.0, in case Zigbee2MQTT cannot automatically detect your adapter, you have to explicitly specify the `serial` `port` and `adapter`.

This change in behaviour has 1 drawback. In case Zigbee2MQTT defaulted to `zstack` before in your setup, it will now fail to start with `USB adapter discovery error (No valid USB adapter found). Specify valid 'adapter' and 'port' in your configuration.`. This can be fixed by adding `adapter: zstack` to your Zigbee2MQTT `configuration.yaml` as described above.

The full documentation can be found [here](https://github.com/Koenkk/zigbee2mqtt.io/blob/fix/adapter-discovery/docs/guide/configuration/adapter-settings.md#basic-configuration).

**Comment by Enzokot** [comment]
Could this error be related to the functionality under discussion? After adding the `adapter: zstack` option the error remains
```
[2024-10-02 09:02:44] error: 	zh:zstack:znp: Failed to determine if path is valid: 'Error: spawn udevadm ENOENT'
[2024-10-02 09:02:44] debug: 	zh:adapter: Failed to validate path: 'Error: spawn udevadm ENOENT'
```
Debug log:
[log.log](https://github.com/user-attachments/files/17241419/log.log)

**Comment by aivult** [comment]
> Could this error be related to the functionality under discussion? After adding the `adapter: zstack` option the error remains
> 
> ```
> [2024-10-02 09:02:44] error: 	zh:zstack:znp: Failed to determine if path is valid: 'Error: spawn udevadm ENOENT'
> [2024-10-02 09:02:44] debug: 	zh:adapter: Failed to validate path: 'Error: spawn udevadm ENOENT'
> ```
> 
> Debug log: [log.log](https://github.com/user-attachments/files/17241419/log.log)

Have the same problem. Adding adapter line doesn't help

**Comment by Koenkk** [comment]
The change has not been merged yet. `Failed to validate path: 'Error: spawn udevadm ENOENT'` means z2m doesn't have udev permissions (for the addon it will be fixed in https://github.com/zigbee2mqtt/hassio-zigbee2mqtt/pull/637)

**Reply by alexsaas** [reply]
Hello @Koenkk I am faced with the same problem and another. Since version 1.40.2-1, I have two error messages in the log of z2m with every complete HA restart, or when I stop z2m and start it again. One before the start and one after:

```
[2024-10-14 03:00:13] error: z2m : MQTT error: read ECONNRESET
[03:01:37] INFO: Preparing to start...
[03:01:37] INFO: Socat not enabled
[03:01:371 INFO: Starting Zigbee2MQTT…
Starting Zigbee2MQTT without watchdog.
[2024-10-14 03:01:39] error: zh: zstack: znp: Failed to determine if path is valid: 'Error: spawn udevadm ENOENT'
```

**Do both error messages match the problem discussed here, or is the first one (when stopping zigbee2MQTT) due to something else?**
Both MQTT and zigbee2MQTT work, but who knows what else these errors can lead to.

**Reply by Koenkk** [reply]
If you get `Error: spawn udevadm ENOENT` but z2m starts fine, there is no need to worry. udev is used for adapter discovery, but if z2m starts you already configured the adapter correctly.

**Reply by alexsaas** [reply]
> If you get `Error: spawn udevadm ENOENT` but z2m starts fine, there is no need to worry. udev is used for adapter discovery, but if z2m starts you already configured the adapter correctly.

Hello @Koenkk thank you very much for your reassuring answer. Can you also tell me something about the first error? 

MQTT error: read ECONNRESET

Is it possible that MQTT cannot be accessed when stopping zigbee2mqtt or Home Assistant? Maybe the broker or the integration stops faster when Home Assistant is shut down than zigbee2MQTT? Should I try to reconfigure the MQTT Mosquitto broker?

I have actually had this error only since the update to version 1.40.2-1. Previously, I had never encountered it for almost a year.

**Comment by popy2k14** [comment]
Same here, thx for the fix.

**Comment by BoxedBrain** [comment]
@Koenkk, when is a new release, including this PR, expected? I am facing the same error, and therefore I am unable to start the service. Thanks

**Comment by 82timo** [comment]
> **TL;DR**
> 
> Add `adapter` (likely `zstack`, see [docs](https://github.com/Koenkk/zigbee2mqtt.io/blob/fix/adapter-discovery/docs/guide/configuration/adapter-settings.md#basic-configuration) for more info) to your Zigbee2MQTT `configuration.yaml`, 


> 
> The full documentation can be found [here](https://github.com/Koenkk/zigbee2mqtt.io/blob/fix/adapter-discovery/docs/guide/configuration/adapter-settings.md#basic-configuration).



Both links are broken:
404 - page not found
Cannot find a valid ref in 
fix/adapter-discovery/docs/guide/configuration/adapter-settings.md

What do I have to do?
In the Addonm configuration tab, serial or in configuration,yaml?

I have a sonoff Zigbee 3.0 USB Dongle Plus

**Comment by Koenkk** [comment]
I will close this discussion and create a new one. It's intended for after the 2.0.0 release (not released yet) but has now turned into a 1.x.x support thread, please open issues for that instead.
